Ability to ship and update card subnets

salsadeanguila opened the pull request:
### ship_card

This changes allows us to call `/v3.1/users/user_id/nodes/node_id/subnets/subnet_id/ship` (`/ship` at the end, different from simple update) with a `PATCH` request as mentioned [here](https://docs.synapsefi.com/reference#native-ship-card).

We are passing the `fee_node_id` which is our `deposit` account ID and the `cardholder_name`.
`expedite` is hard coded as `false` and `card_style_id` is not being sent so Synapse will use our default card style.

### subnet#update

Allows us to update `card subnets` (account_class: `CARD`). This is used for activation/termination of cards.

salsadeanguila committed:
ship path builder

We need to call that route with a patch to provision the card

salsadeanguila committed:
ship_card method to provision card

salsadeanguila committed:
Handle response for ship_card method

salsadeanguila committed:
Fixed return comment

salsadeanguila committed:
update subnet instance method and payload creator for card subnet update

salsadeanguila committed:
Added missing params

salsadeanguila committed:
Remove unnecessary params

salsadeanguila committed:
Add missing calls

salsadeanguila committed:
Call method correctly

salsadeanguila committed:
payload_for_card_update private instance method instead of private class method

salsadeanguila committed:
Updated payload_for_card_update

safe navigation for preferences

salsadeanguila committed:
Pass node instead of client

fertech commented on lib/synapse_pay_rest/models/subnet/subnet.rb:
```diff
@@ -174,9 +174,121 @@ def lock
+    # Converts #update args into API payload structure for CARD subnets.
+    def payload_for_card_update(**options)
+      payload = {}
+      # must have one of these
```
do we need to raise an error if at least one is not present?

fertech commented on lib/synapse_pay_rest/models/subnet/subnet.rb:
```diff
@@ -174,9 +174,121 @@ def lock
+
+      if options[:preferences]
+        if options[:preferences][:allow_foreign_transactions]
+          payload['preferences']['allow_foreign_transactions']  = options[:preferences][:allow_foreign_transactions]
```
are you sure that we can add `payload['preferences']['allow_foreign_transactions'] ` when `payload['preferences']` has not been defined as a key yet? In the console, I first need to do `payload['preferences'] = {}`.

Also, not sure if the DRY'ness is worth the readability, but you could do something like:

    if options[:preferences] 
        options[:preferences].each do |key ,value| do
           payload['preferences'][key.to_s] = value
       end
    end

lukeasrodgers replied:
agree

fertech reviewed (changes requested):
Looks great!
A possible bug with `payload['preferences']`. 
Would be good if you could add tests, not sure how we go about it in this forked branch.

salsadeanguila committed:
Addressed comments

Moved the raising of the error to the payload method, declared "preferences" key, refactored way to extract keys from options array and changed card_pin to pin because of Keven's request

lukeasrodgers reviewed (approved):
👍
